File: app/channels/participant_presence_channel.rb
class ParticipantPresenceChannel < ApplicationCable::Channel
  def subscribed
  end

  def unsubscribed
    leave_game
  end

  def arrive_at_game(data)
    if data['game_id']
      game = Game.find(data['game_id'])
      if game && connection.current_user
        connection.current_user.arrive_at_game(game)
      end
    end
  end

  def leave_game
    if connection.current_user
      connection.current_user.leave_game
    end
  end
end

File: app/javascript/channels/participant_presence_channel.js
import consumer from "channels/consumer"

export default consumer.subscriptions.create("ParticipantPresenceChannel", {
  connected() {
    // Called when the subscription is ready for use on the server
  },

  disconnected() {
    this.leaveGame()
  },

  arriveAtGame(gameId) {
    this.perform('arrive_at_game', {game_id: gameId})
  },

  leaveGame() {
    this.perform('leave_game')
  },

  received(data) {
    // Called when there's incoming data on the websocket for this channel
  }
});
